zebra-chain/src/amount.rs

```diff
@@ -191,6 +191,54 @@ where
     }
 }
 
+impl<C> std::ops::Mul<i64> for Amount<C>
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
+
+    fn mul(self, rhs: i64) -> Self::Output {
+        let value = self.0 * rhs;
+        value.try_into()
+    }
+}
+
+impl<C> std::ops::Mul<Amount<C>> for i64
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
+
+    fn mul(self, rhs: Amount<C>) -> Self::Output {
+        let value = rhs.0 * self;
+        value.try_into()
+    }
+}
+
+impl<C> std::ops::Div<i64> for Amount<C>
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
+
+    fn div(self, rhs: i64) -> Self::Output {
+        let value = self.0 / rhs; // Todo: need division errors
+        value.try_into()
+    }
+}
```

Comment:
Let's handle divide by zero as an error, and truncate the remainder. (Like integer division.)

(Since `Amount` is constrained to `-MAX_MONEY..=MAX_MONEY`, it is not possible for `self / -1` to overflow. This is different from unconstrained integers,  where`i64::MIN / -1` will overflow.)

Something like this should work - but I haven't checked if it compiles:

```suggestion
#[derive(thiserror::Error, Debug, displaydoc::Display, Clone, PartialEq)]
#[allow(missing_docs)]
enum AmountDivError<C> {
    /// division by zero is an invalid operation, input {value}
    DivideByZero {
        value: i64,
    },
};

impl<C> std::ops::Div<u64> for Amount<C>
where
    C: Constraint,
{
    type Output = Result<Amount<C>, AmountDivError<C>>;

    fn div(self, rhs: u64) -> Self::Output {
        let quotient = self.0.checked_div(rhs).ok_or(DivideByZero)?;
        // since this is a division by a positive integer,
        // the quotient must be within the constrained range
        Ok(quotient.try_into().unwrap())
    }
}
```

Reply:
Actually, I think it's enough to round down the result.

We can document this behaviour by leaving comments in the places we use Amount division - there are currently 2 places:
- [x] block subsidy halving - exact now, round down in future halvings
- [x] founders reward - exact

And we will add one more for the founders reward:
- [ ] major grants, ecc, and zcash foundation slices - inexact for ecc now (`/7`) and everyone later (`/8` after 4 halvings?, `/5` after 7 halvings??)